zebra-chain/src/primitives/redpallas/tests/batch.rs

```diff
@@ -0,0 +1,99 @@
+use rand::thread_rng;
+
+use super::super::*;
+
+#[test]
+fn spendauth_batch_verify() {
+    let mut rng = thread_rng();
+    let mut batch = batch::Verifier::new();
+    for _ in 0..32 {
+        let sk = SigningKey::<SpendAuth>::new(&mut rng);
+        let vk = VerificationKey::from(&sk);
+        let msg = b"BatchVerifyTest";
+        let sig = sk.sign(&mut rng, &msg[..]);
+        batch.queue((vk.into(), sig, msg));
+    }
+    assert!(batch.verify(rng).is_ok());
+}
+
+#[test]
+fn binding_batch_verify() {
+    let mut rng = thread_rng();
+    let mut batch = batch::Verifier::new();
+    for _ in 0..32 {
+        let sk = SigningKey::<Binding>::new(&mut rng);
+        let vk = VerificationKey::from(&sk);
+        let msg = b"BatchVerifyTest";
+        let sig = sk.sign(&mut rng, &msg[..]);
+        batch.queue((vk.into(), sig, msg));
+    }
+    assert!(batch.verify(rng).is_ok());
+}
+
+#[test]
+fn alternating_batch_verify() {
+    let mut rng = thread_rng();
+    let mut batch = batch::Verifier::new();
+    for i in 0..32 {
+        let item: batch::Item = match i % 2 {
+            0 => {
+                let sk = SigningKey::<SpendAuth>::new(&mut rng);
+                let vk = VerificationKey::from(&sk);
+                let msg = b"BatchVerifyTest";
+                let sig = sk.sign(&mut rng, &msg[..]);
+                (vk.into(), sig, msg).into()
+            }
+            1 => {
+                let sk = SigningKey::<Binding>::new(&mut rng);
+                let vk = VerificationKey::from(&sk);
+                let msg = b"BatchVerifyTest";
+                let sig = sk.sign(&mut rng, &msg[..]);
+                (vk.into(), sig, msg).into()
+            }
+            _ => unreachable!(),
+        };
+        batch.queue(item);
+    }
+    assert!(batch.verify(rng).is_ok());
+}
+
+#[test]
+fn bad_batch_verify() {
+    let mut rng = thread_rng();
+    let bad_index = 4; // must be even
```

Comment:
These tests all look really good.

It might also be worth testing a bad `Binding` signature, because we've had binding-signature specific bugs before.

(You could turn this into a proptest that takes one or more bad signature indexes.)

Reply:
Done via a70cd457